### PR TITLE
New offer feature

### PR DESCRIPTION
**Pull request by psorinionut**
- Added the offer functionality
- Added an extra functionality that ends an auction when maxBid has been reached
- Updated tests & code refactors

### CODE REVIEW

File: esdt-nft-marketplace/src/offer.rs [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+            .swap_remove(&offer_id);
+        self.offer_by_id(offer_id).clear();
+
+        self.emit_withdraw_offer_event(offer_id, offer);
+    }
+
+    #[payable("*")]
+    #[endpoint(acceptOffer)]
+    fn accept_offer(&self, offer_id: u64) {
+        let offer_nft = self.call_value().single_esdt();
```
Comment (CostinCarabas) [inline esdt-nft-marketplace/src/offer.rs:104]: `self.call_value().single_esdt();` -> `self.call_value().egld_or_single_esdt();` ?

File: esdt-nft-marketplace/src/offer.rs [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+        require!(
+            offer_nft.token_identifier == offer.offer_token.token_identifier,
+            "The sent token type is different from the offer"
+        );
+        require!(
+            offer_nft.token_nonce == offer.offer_token.token_nonce,
+            "The sent token nonce is different from the offer"
+        );
+        require!(
+            offer_nft.amount == 1u64,
```
Comment (CostinCarabas) [inline esdt-nft-marketplace/src/offer.rs:120]: Maybe use `NFT_AMOUNT`

File: esdt-nft-marketplace/src/offer.rs (hunk 1 above)
Comment (dorin-iancu) [reply esdt-nft-marketplace/src/offer.rs:104]: Nope, the caller is expected to deposit the NFT for which another user made the offer.

Comment (psorinionut) [reply esdt-nft-marketplace/src/offer.rs:104]: acceptOffer means you need to send the NFT to the contract, so the payment must be only of type ESDT. Also, `single_esdt()` returns `EsdtTokenPayment` which makes it easier to pass it along.

File: esdt-nft-marketplace/src/offer.rs (hunk 2 above)
Comment (psorinionut) [reply esdt-nft-marketplace/src/offer.rs:120]: Done.

File: esdt-nft-marketplace/src/offer.rs [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+        let offer = Offer {
+            offer_token,
+            payment,
+            start_time: current_time,
+            deadline,
+            offer_owner: caller,
+        };
+
+        let offer_id = self.last_valid_offer_id().get() + 1;
```
Comment (sasurobert) [inline esdt-nft-marketplace/src/offer.rs:57]: add check if the NFT is in the marketplace and has an ongoing auction with at least one bid then transaction should fail.

**Review by sasurobert — COMMENTED** [review]
offer endpoints not complete

File: esdt-nft-marketplace/src/offer.rs (hunk 3 above)
Comment (psorinionut) [reply esdt-nft-marketplace/src/offer.rs:82]: Done.

File: esdt-nft-marketplace/src/offer.rs [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+    }
+
+    #[payable("*")]
+    #[endpoint(acceptOffer)]
+    fn accept_offer(&self, offer_id: u64) {
+        let offer_nft = self.call_value().single_esdt();
+        let offer = self.try_get_offer(offer_id);
+        let seller = self.blockchain().get_caller();
+        let current_time = self.blockchain().get_block_timestamp();
+        require!(current_time < offer.deadline, "Offer has expired");
```
[the remark this answers is not in the document]
Comment (psorinionut) [reply esdt-nft-marketplace/src/offer.rs]: Done. Added an extra endpoint to accept other input that the actual NFT as a payment.

**Review by sasurobert — DISMISSED** [review]

File: esdt-nft-marketplace/src/offer.rs [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+
+        let offer = Offer {
+            offer_token,
+            payment,
+            start_time: current_time,
+            deadline,
+            offer_owner: caller,
+        };
+
+        let token_amount_in_marketplace = self.blockchain().get_sc_balance(
```
Comment (sasurobert) [inline esdt-nft-marketplace/src/offer.rs:61]: move this into a separate function.

File: esdt-nft-marketplace/src/offer.rs [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+                    );
+                }
+                OptionalValue::None => sc_panic!("Must provide the auction id"),
+            };
+        }
+
+        let offer_id = self.last_valid_offer_id().get() + 1;
+        self.last_valid_offer_id().set(&offer_id);
+        self.offer_by_id(offer_id).set(&offer);
+        self.offers_by_address(&offer.offer_owner).insert(offer_id);
```
Comment (sasurobert) [inline esdt-nft-marketplace/src/offer.rs:85]: no need.

File: esdt-nft-marketplace/src/offer.rs [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+        self.offers_by_token(
```
Comment (sasurobert) [inline esdt-nft-marketplace/src/offer.rs:86]: no need.

File: esdt-nft-marketplace/src/offer.rs [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+            offer.payment.token_nonce,
+            &offer.payment.amount,
+        );
+
+        self.offers_by_token(
+            &offer.offer_token.token_identifier,
+            offer.offer_token.token_nonce,
+        )
+        .swap_remove(&offer_id);
+        self.offers_by_address(&offer.offer_owner)
```
Comment (sasurobert) [inline esdt-nft-marketplace/src/offer.rs:120]: delete these storrages.

File: esdt-nft-marketplace/src/offer.rs (hunk 5 above)
Comment (psorinionut) [reply esdt-nft-marketplace/src/offer.rs:75]: Done.

File: esdt-nft-marketplace/src/offer.rs (hunk 6 above)
Comment (psorinionut) [reply esdt-nft-marketplace/src/offer.rs]: Done.

File: esdt-nft-marketplace/src/offer.rs (hunk 7 above)
Comment (psorinionut) [reply esdt-nft-marketplace/src/offer.rs]: Done.

File: esdt-nft-marketplace/src/offer.rs (hunk 8 above)
Comment (psorinionut) [reply esdt-nft-marketplace/src/offer.rs]: Done.

**Review by sasurobert — APPROVED** [review]

**Review by sasurobert — DISMISSED** [review]

**Review by sasurobert — APPROVED** [review]

**Pull request merged**